File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-6.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        ref= head
        for i in range(n):
            ref = ref.next
        cur = head
        if not ref:
            return head.next
        while ref.next:
            ref = ref.next
            cur = cur.next
        cur.next = cur.next.next
        return head
        

File: Data Structures & Algorithms/reorder-linked-list/submission-6.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        while(fast != null && fast.next != null){
            fast = fast.next.next;
            slow = slow.next;
        }
        ListNode second = slow.next;
        ListNode prev = slow.next = null;   
        while(second != null){
            ListNode temp = second.next;
            second.next = prev;
            prev = second;
            second = temp;
        }
        second = prev;
        ListNode first = head;
        while(second != null){
            ListNode firstTemp = first.next;
            ListNode secondTemp = second.next;
            first.next = second;
            second.next = firstTemp;
            first = firstTemp;second = secondTemp;
        }
    }
}
